### ISSUE TITLE
False Positive on while_let_on_iterator

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust-clippy#1121]
I just ran into a false positive with while_let_on_iterator. I have a HashSet and as long as it contains elements, I want to do some work and maybe insert some more elements. This wouldn't work with a for loop as it borrows the whole collection for the scope of the loop. So I'm using while let in a non-borrowing way instead, which works perfectly fine. However clippy suggests a for loop instead which breaks the code.

[Playground Link](https://is.gd/SG0zr7)

``` rust
use std::collections::HashSet;

fn main() {
    let mut values = HashSet::new();
    values.insert(1);

    // Original Code:
    while let Some(&value) = values.iter().next() {
        values.remove(&value);
        // Do stuff that inserts values
    }

    // Clippy Suggestion (doesn't compile):
    for &value in &values {
        values.remove(&value);
        // Do stuff that inserts values
    }
}
```

**Comment by mikerite** [comment]
I can't reproduce the bad suggestion. 
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=449489da2835027264a57d5b48ecfdb6)

I think we fixed this one.
